napari/_qt/qt_performance.py

```diff
@@ -0,0 +1,174 @@
+"""QtPerformance widget to show performance information.
+"""
+import time
+
+from qtpy.QtCore import QTimer, Qt
+from qtpy.QtGui import QTextCursor
+
+from qtpy.QtWidgets import (
+    QComboBox,
+    QHBoxLayout,
+    QLabel,
+    QProgressBar,
+    QSizePolicy,
+    QSpacerItem,
+    QTextEdit,
+    QVBoxLayout,
+    QWidget,
+)
+
+from ..utils import perf
+
+
+class TextLog(QTextEdit):
+    """Text window we can write "log" messages to.
+
+    TODO: need to limit length, erase oldest messages.
+    """
+
+    def append(self, name, time_ms):
+        """Add one line of text for this timer.
+        """
+        self.moveCursor(QTextCursor.End)
+        self.setTextColor(Qt.red)
+        self.insertPlainText(f"{time_ms:5.0f}ms {name}\n")
+
+
+class QtPerformance(QWidget):
+    """Dockable widget to show performance info.
+
+    Notes
+    -----
+
+    1) The progress bar doesn't show "progress", we use it as a bar graph to
+       show the average duration of recent "UpdateRequest" events. This
+       is actually not the total draw time, but it's generally the biggest
+       part of each frame.
+
+    2) We log any event whose duration is longer than the threshold.
+
+    3) We show uptime so you can tell if this window is being updated at all.
+    """
+
+    # We log events slower than some threshold (in milliseconds).
+    THRESH_DEFAULT = 100
+    THRESH_OPTIONS = [
+        "1",
+        "5",
+        "10",
+        "15",
+        "20",
+        "30",
+        "40",
+        "50",
+        "100",
+        "200",
+    ]
+
+    # Update at 250ms / 4Hz for now. The more we update more alive our
+    # display will look, but the more we will slow things down.
+    UPDATE_MS = 250
+
+    def __init__(self):
+        """Create our windgets.
+        """
+        super().__init__()
+        layout = QVBoxLayout()
+
+        # For our "uptime" timer.
+        self.start_time = time.time()
+
+        # Label for our progress bar.
+        bar_label = QLabel("Draw Time:")
+        layout.addWidget(bar_label)
+
+        # Progress bar is not used for "progress", it's just a bar graph to show
+        # the "draw time", the duration of the "UpdateRequest" event.
+        bar = QProgressBar()
+        bar.setRange(0, 100)
+        bar.setValue(50)
+        bar.setFormat("%vms")
+        layout.addWidget(bar)
+        self.bar = bar
+
+        # We let the user set the "slow event" threshold.
+        self.thresh_ms = self.THRESH_DEFAULT
+        self.thresh_combo = QComboBox()
+        self.thresh_combo.addItems(self.THRESH_OPTIONS)
```

Comment:
another trick worth knowing about... the [`QComboBox.addItem` method](https://doc.qt.io/qt-5/qcombobox.html#addItem) (but not `addItems`) accepts an optional _UserData_ value that lets you store arbitrary data with an item ... which you can then retrieve using [`currentData`](https://doc.qt.io/qt-5/qcombobox.html#currentData-prop).  This let's you avoid all the boilerplate of needing to convert back and forth between strings and floats

Reply:
That's good to know about. This doesn't seem bad with string/float but I could see `UserData` being cleaner in some cases.